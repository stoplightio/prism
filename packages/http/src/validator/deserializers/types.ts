import { HttpParamStyles } from '@stoplight/types';
import { IHttpNameValue, IHttpNameValues, JSONSchema } from '../../types';

export type deserializeFn<Parameters> = (
  name: string,
  parameters: Parameters,
  schema?: JSONSchema,
  explode?: boolean
) => unknown;

export interface IHttpParamDeserializerRegistry<Parameters, S = HttpParamStyles> {
  get(style: S): IHttpParamStyleDeserializer<Parameters> | undefined;
}

export interface IHttpParamStyleDeserializer<Parameters, S = HttpParamStyles> {
  supports: (style: S) => boolean;
  deserialize: deserializeFn<Parameters>;
}

export type IHttpHeaderParamStyleDeserializer = IHttpParamStyleDeserializer<IHttpNameValue>;
export type IHttpQueryParamStyleDeserializer = IHttpParamStyleDeserializer<IHttpNameValues>;
