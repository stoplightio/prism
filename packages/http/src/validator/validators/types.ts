import { IPrismDiagnostic } from '@stoplight/prism-core/src/types';
import { JSONSchema } from 'http/src/types';

export interface IHttpValidator<Target, Specs> {
  validate(target: Target, specs: Specs[], mediaType?: string): IPrismDiagnostic[];
}

export interface ISchemaValidator<S = JSONSchema> {
  validate(content: any, schema: S): IPrismDiagnostic[];
  supports(mediaType: string): boolean;
}

export interface IValidatorRegistry {
  get(mediaType: string): ((content: any, schema: JSONSchema) => IPrismDiagnostic[]) | undefined;
}
