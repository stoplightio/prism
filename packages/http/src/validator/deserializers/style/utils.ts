import { Dictionary } from '@stoplight/types';

export function createObjectFromKeyValList(items: string[]) {
  return items.reduce((obj: Dictionary<unknown>, item, i) => {
    if (i % 2 === 0) {
      obj[item] = undefined;
    } else {
      obj[items[i - 1]] = item;
    }

    return obj;
  }, {});
}
