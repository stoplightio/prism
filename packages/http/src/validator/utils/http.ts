import { IHttpContent } from '@stoplight/types';

export function resolveContent(content: { [mediaType: string]: IHttpContent }, mediaType?: string) {
  return content[mediaType || '*'] || content['*'];
}
