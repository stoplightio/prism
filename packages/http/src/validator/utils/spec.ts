import { IHttpOperationResponse } from '@stoplight/types';
import { head } from 'fp-ts/lib/Array';
import { Option } from 'fp-ts/lib/Option';
import { NonEmptyArray } from 'fp-ts/lib/NonEmptyArray';

export function findOperationResponse(
  responseSpecs: NonEmptyArray<IHttpOperationResponse>,
  statusCode: number
): Option<IHttpOperationResponse> {
  const sortedSpecs = responseSpecs
    .filter(
      spec => new RegExp(`^${spec.code.replace(/X/g, '\\d')}$`).test(String(statusCode)) || spec.code === 'default'
    )
    .sort((s1, s2) => {
      if (s1.code === 'default') {
        return 1;
      }

      if (s2.code === 'default') {
        return -1;
      }

      return s1.code.split('X').length - s2.code.split('X').length;
    });

  return head(sortedSpecs);
}
