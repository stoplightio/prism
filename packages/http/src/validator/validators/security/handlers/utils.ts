import { DiagnosticSeverity } from '@stoplight/types';
import { Either, left, right } from 'fp-ts/lib/Either';
import { IPrismDiagnostic } from '@stoplight/prism-core';
import { IHttpRequest } from '../../../../types';

export type ValidateSecurityFn = (
  input: Pick<IHttpRequest, 'headers' | 'url'>,
  name: string
) => Either<IPrismDiagnostic, unknown>;

export function genRespForScheme(
  isSchemeProper: boolean,
  isCredsGiven: boolean,
  msg: string
): Either<IPrismDiagnostic, unknown> {
  if (isSchemeProper) {
    return when(isCredsGiven, undefined);
  }

  return left(genUnauthorisedErr(msg));
}

export const genUnauthorisedErr = (msg?: string): IPrismDiagnostic => ({
  severity: DiagnosticSeverity.Error,
  message: 'Invalid security scheme used',
  code: 401,
  tags: msg ? [msg] : [],
});

export function isScheme(shouldBeScheme: string, authScheme: string) {
  return authScheme.toLowerCase() === shouldBeScheme;
}

export function when(condition: boolean, errorMessage: string | undefined): Either<IPrismDiagnostic, unknown> {
  return condition ? right(true) : left(genUnauthorisedErr(errorMessage));
}
