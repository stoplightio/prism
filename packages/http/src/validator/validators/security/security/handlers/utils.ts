import { DiagnosticSeverity, IHttpOperation } from '@stoplight/types';
import { Either, left, right } from 'fp-ts/lib/Either';
import { IPrismDiagnostic } from '@stoplight/prism-core';

export function genRespForScheme(
  isSchemeProper: boolean,
  isCredsGiven: boolean,
  resource: IHttpOperation,
  msg: string
): Either<IPrismDiagnostic, IHttpOperation> {
  if (isSchemeProper) {
    return when(isCredsGiven, undefined, resource);
  }

  return left(genUnauthorisedErr(msg));
}

export const genUnauthorisedErr = (msg?: string): IPrismDiagnostic => ({
  severity: DiagnosticSeverity.Error,
  message: 'Invalid security scheme used',
  code: 401,
  tags: msg ? [msg] : [],
});

export function isScheme(shouldBeScheme: string, authScheme: string) {
  return authScheme.toLowerCase() === shouldBeScheme;
}

export function when(
  condition: boolean,
  errorMessage: string | undefined,
  resource: IHttpOperation
): Either<IPrismDiagnostic, IHttpOperation> {
  return condition ? right(resource) : left(genUnauthorisedErr(errorMessage));
}
