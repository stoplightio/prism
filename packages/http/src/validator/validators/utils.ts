import { IPrismDiagnostic } from '@stoplight/prism-core';
import { DiagnosticSeverity, Segment } from '@stoplight/types';
import { getSemigroup } from 'fp-ts/lib/NonEmptyArray';
import { getValidation } from 'fp-ts/lib/Either';
import { option, tryCatch, Option } from 'fp-ts/lib/Option';
import { sequenceT } from 'fp-ts/lib/Apply';
import * as Ajv from 'ajv';
import { JSONSchema } from '../../';
import * as AjvOAI from 'ajv-oai';

const ajv = new AjvOAI({ allErrors: true, messages: true, schemaId: 'auto' });

export const convertAjvErrors = (errors: Ajv.ErrorObject[] | undefined | null, severity: DiagnosticSeverity) => {
  if (!errors) {
    return [];
  }

  return errors.map<IPrismDiagnostic & { path: Segment[] }>(error => {
    const allowedParameters = 'allowedValues' in error.params ? `: ${error.params.allowedValues.join(', ')}` : '';

    return {
      path: error.dataPath.split('.').slice(1),
      code: error.keyword || '',
      message: `${error.message || ''}${allowedParameters}`,
      severity,
    };
  });
};

export const validateAgainstSchema = (
  value: unknown,
  schema: JSONSchema,
  prefix?: string
): Option<IPrismDiagnostic[]> => {
  return tryCatch(() => {
    const validate = ajv.compile(schema);
    const valid = validate(value);
    if (!valid) {
      return convertAjvErrors(validate.errors, DiagnosticSeverity.Error).map(error => {
        const path = prefix ? [prefix, ...error.path] : error.path;
        return Object.assign({}, error, { path });
      });
    }
    return [];
  });
};

export const sequenceValidation = sequenceT(getValidation(getSemigroup<IPrismDiagnostic>()));
export const sequenceOption = sequenceT(option);
