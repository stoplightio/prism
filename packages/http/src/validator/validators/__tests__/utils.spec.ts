import { DiagnosticSeverity } from '@stoplight/types';
import * as convertAjvErrorsModule from '../utils';
import { convertAjvErrors, validateAgainstSchema } from '../utils';
import { ErrorObject } from 'ajv';
import { assertSome, assertNone } from '@stoplight/prism-core/src/__tests__/utils';

describe('convertAjvErrors()', () => {
  const errorObjectFixture: ErrorObject = {
    dataPath: 'a.b',
    keyword: 'required',
    message: 'c is required',
    schemaPath: '..',
    params: {},
  };

  describe('all fields defined', () => {
    it('converts properly', () => {
      expect(convertAjvErrors([errorObjectFixture], DiagnosticSeverity.Error)).toMatchSnapshot();
    });
  });

  describe('keyword field is missing', () => {
    it('converts properly', () => {
      expect(
        convertAjvErrors([Object.assign({}, errorObjectFixture, { keyword: undefined })], DiagnosticSeverity.Error)
      ).toMatchSnapshot();
    });
  });

  describe('message field is missing', () => {
    it('converts properly', () => {
      expect(
        convertAjvErrors([Object.assign({}, errorObjectFixture, { message: undefined })], DiagnosticSeverity.Error)[0]
      ).toHaveProperty('message', '');
    });
  });
});

describe('validateAgainstSchema()', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.spyOn(convertAjvErrorsModule, 'convertAjvErrors');
  });

  describe('has no validation errors', () => {
    it('returns no validation errors', () => {
      assertNone(validateAgainstSchema('test', { type: 'string' }, 'pfx'));
      expect(convertAjvErrorsModule.convertAjvErrors).not.toHaveBeenCalled();
    });
  });

  describe('has validation errors', () => {
    it('returns validation errors', () => {
      jest.spyOn(convertAjvErrorsModule, 'convertAjvErrors').mockImplementationOnce(() => [
        {
          message: 'should be number',
          code: '10',
          path: [],
          severity: DiagnosticSeverity.Error,
          summary: 'should be number',
        },
      ]);
      assertSome(validateAgainstSchema('test', { type: 'number' }, 'pfx'), error =>
        expect(error).toContainEqual(expect.objectContaining({ message: 'should be number' }))
      );

      expect(convertAjvErrorsModule.convertAjvErrors).toHaveBeenCalledWith(
        [
          {
            dataPath: '',
            keyword: 'type',
            message: 'should be number',
            params: { type: 'number' },
            schemaPath: '#/type',
          },
        ],
        DiagnosticSeverity.Error,
        'pfx'
      );
    });
  });
});
