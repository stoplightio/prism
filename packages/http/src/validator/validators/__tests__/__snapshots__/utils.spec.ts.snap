// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertAjvErrors() all fields defined converts properly 1`] = `
Array [
  Object {
    "code": "required",
    "message": "c is required",
    "path": Array [
      "b",
    ],
    "severity": 0,
  },
]
`;

exports[`convertAjvErrors() keyword field is missing converts properly 1`] = `
Array [
  Object {
    "code": "",
    "message": "c is required",
    "path": Array [
      "b",
    ],
    "severity": 0,
  },
]
`;
